src/terminal/parser/stateMachine.cpp

```diff
@@ -1270,6 +1436,220 @@ void StateMachine::_EventVt52Param(const wchar_t wch)
+    {
+        _ActionIgnore();
+    }
+    else if (_isIntermediate(wch))
+    {
+        _ActionCollect(wch);
+    }
+    else if (_isIntermediateInvalid(wch))
+    {
+        _EnterDcsIgnore();
+    }
+    else
+    {
+        _ActionDcsPassThrough(wch);
+        _EnterDcsPassThrough();
+    }
+}
+
+// Routine Description:
+// - Processes a character event into an Action that occurs while in the DcsParam state.
+//   Events in this state will:
+//   1. Ignore C0 control characters
+//   2. Ignore Delete characters
+//   3. Collect DCS parameter data
+//   4. Enter DcsIntermediate if we see an intermediate
+//   5. Begin to ignore all remaining parameters when an invalid character is detected (DcsIgnore)
+//   6. Pass through everything else.
+// Arguments:
+// - wch - Character that triggered the event
+// Return Value:
+// - <none>
+void StateMachine::_EventDcsParam(const wchar_t wch)
+{
+    _trace.TraceOnEvent(L"DcsParam");
+    if (_isC0Code(wch))
+    {
+        _ActionIgnore();
+    }
+    else if (_isDelete(wch))
+    {
+        _ActionIgnore();
+    }
+    if (_isNumericParamValue(wch) || _isParameterDelimiter(wch))
+    {
+        _ActionParam(wch);
+    }
+    else if (_isIntermediate(wch))
+    {
+        _ActionCollect(wch);
+        _EnterDcsIntermediate();
+    }
+    else if (_isParameterInvalid(wch))
+    {
+        _EnterDcsIgnore();
+    }
+    else
+    {
+        _ActionDcsPassThrough(wch);
+        _EnterDcsPassThrough();
+    }
+}
+
+// Routine Description:
+// - Processes a character event into an Action that occurs while in the DcsPassThrough state.
+//   Events in this state will:
+//   1. Enter ground on a String terminator
+//   2. Pass through if character is valid.
+//   3. If we see a ESC, enter the DcsTermination state.
+//   4. Ignore everything else.
+// Arguments:
+// - wch - Character that triggered the event
+// Return Value:
+// - <none>
+void StateMachine::_EventDcsPassThrough(const wchar_t wch)
+{
+    _trace.TraceOnEvent(L"DcsPassThrough");
+    if (_isStringTerminator(wch))
+    {
+        _EnterGround();
+    }
+    if (_isC0Code(wch) || _isDcsPassThroughValid(wch))
+    {
+        _ActionDcsPassThrough(wch);
+    }
+    else if (_isDcsTerminationInitiator(wch))
+    {
+        _EnterDcsTermination();
+    }
+    else
+    {
+        _ActionIgnore();
+    }
+}
+
+// Routine Description:
+// - Handle the two-character termination of a DCS sequence.
+//   Events in this state will:
+//   1. Enter ground on a string terminator
+//   2. Pass on everything else as the start of a regular escape sequence
+// Arguments:
+// - wch - Character that triggered the event
+// Return Value:
+// - <none>
+void StateMachine::_EventDcsTermination(const wchar_t wch) noexcept
+{
+    _trace.TraceOnEvent(L"DcsTermination");
+
+    if (wch == L'\\')
+    {
+        // TODO: This is where we'd be dispatching the DCS command.
+        _EnterGround();
+    }
+    else
+    {
+        _EventEscape(wch);

```

Comment:
I just realised that we probably also need a call to `_EnterEscape` here, before the call to `_EventEscape`. If the escape wasn't followed by `\`, then we're essentially cancelling the current sequence and started a new one, and we need to reset any parameters and intermediates that we may already have accumulated.

Reply:
It occurred to me that the `OSC` sequences should be doing the same thing as this, but I see that `_EventOscTermination`, doesn't care what follows the escape - it just dispatches the sequence anyway. That's definitely a bug. Not that that's your problem - just noting that we need to raise an issue for that.

Reply:
Yeah I observed a lot of similarity between OSC and DCS. Perhaps the logic of those two should be more unified in some way. I created `_isIntermediateInvalid` and `_isParameterInvalid` to make things a little bit generic than before.

Reply:
I think that'll become more important if/when we do APC and PM, but DCS has a few differences from the others, so it may not be as easy to share things. What you've done so far with the generic methods is at least a step in the right direction, and we can always do more refactoring later if necessary.

Reply:
By the way why don't we support C1Osc? Is there a reason or it should be allowed to enter from ground?

Reply:
I haven't looked at this side of things in much detail, and I think there is some debate about the use of C1 controls with UTF-8, but if we're supporting some of them, then I would have thought we should be supporting all of them (and that includes the likes of `IND`, `NEL`, `HTS`, etc.).

I was actually wondering whether this was something we could handle as a sort of preprocessing step that converts all of the C1 controls into their two character 7-bit equivalent. That way we can drop all the extra code that's checking for 90 and 9B introducers, and 9C terminators, because those conditions would now be handled as regular 7-bit escapes.

Reply:
Yeah that's a good idea, to convert C1 controls to their 7 bit equivalent. I currently do not have enough experience with VT to file a PR but I'm getting there. Have learned a lot throughout the review period of this PR.

Reply:
Chester, would you mind turning this thread into a followup issue? It seems like we should fix OSC Termination to handle ESC more correctly.

Reply:
Yeah we should. I plan to do that next. Filed #7317 for this.

Reply:
Yeah we should. I plan to do that next. Filed #7317 for this.